### PR TITLE
Refs #15851 - support for dhcp and dns via infoblox

### PR DESCRIPTION
**Pull request by iNecas**

### CODE REVIEW

File: manifests/plugin/dhcp/infoblox.pp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# $record_type:: Record type to manage, can be "host" or "fixedaddress"
+#
+# $use_ranges::  Use  pre-definded ranges in networks to find available IP's
+#                type:boolean
+#
+class foreman_proxy::plugin::dhcp::infoblox (
+  $username    = $::foreman_proxy::plugin::dhcp::infoblox::params::username,
+  $password    = $::foreman_proxy::plugin::dhcp::infoblox::params::password,
+  $record_type = $::foreman_proxy::plugin::dhcp::infoblox::params::record_type,
+  $use_ranges  = $::foreman_proxy::plugin::dhcp::infoblox::params::use_ranges
```
Comment (ekohl) [inline manifests/plugin/dhcp/infoblox.pp]: nit: please add a trailing comma

File: manifests/plugin/dhcp/infoblox/params.pp [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class foreman_proxy::plugin::dhcp::infoblox::params {
+  $username    = undef
+  $password    = undef
+  $record_type = 'fixedaddress'
+  $use_ranges  = false
+}
+
```
Comment (ekohl) [inline manifests/plugin/dhcp/infoblox/params.pp]: nit: can you remove this empty line?

File: manifests/plugin/dns/infoblox.pp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# $dns_server:: The address of the Infoblox server
+#
+# $username::   The username of the Infoblox user
+#
+# $password::   The password of the Infoblox user
+#
+class foreman_proxy::plugin::dns::infoblox (
+  $dns_server = $::foreman_proxy::plugin::dns::infoblox::params::dns_server,
+  $username   = $::foreman_proxy::plugin::dns::infoblox::params::username,
+  $password   = $::foreman_proxy::plugin::dns::infoblox::params::password
```
Comment (ekohl) [inline manifests/plugin/dns/infoblox.pp]: nit: can you add a trailing comma?

File: manifests/plugin/dns/infoblox/params.pp [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class foreman_proxy::plugin::dns::infoblox::params {
+  $dns_server = undef
+  $username   = undef
+  $password   = undef
+}
+
```
Comment (ekohl) [inline manifests/plugin/dns/infoblox/params.pp]: nit: can you remove this empty line?

**Comment by ekohl** [comment]
Code looks good, just some minor things. Could you also add a spec for the dhcp plugin?

**Comment by iNecas** [comment]
@ekohl thanks for review, I've updated the code based on comments: I had the spec for dhcp, but forgot to push it, it's there now.

**Comment by ekohl** [comment]
:+1: on green tests.

**Comment by brandonweeks** [comment]
LGTM. Squash commits?

**Comment by brandonweeks** [comment]
@ekohl perhaps I'm missing something but with how the repository settings are configured they're going to get squashed on merge regardless?

**Comment by mmoll** [comment]
dns and dhcp are independet, so I cherry-picked this as two commits: a9e8d88 (dns) and 5f6393d (dhcp), díky @iNecas!

File: manifests/plugin/dhcp/infoblox.pp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# $use_ranges::  Use  pre-definded ranges in networks to find available IP's
+#                type:boolean
+#
+class foreman_proxy::plugin::dhcp::infoblox (
+  $username    = $::foreman_proxy::plugin::dhcp::infoblox::params::username,
+  $password    = $::foreman_proxy::plugin::dhcp::infoblox::params::password,
+  $record_type = $::foreman_proxy::plugin::dhcp::infoblox::params::record_type,
+  $use_ranges  = $::foreman_proxy::plugin::dhcp::infoblox::params::use_ranges,
+) inherits foreman_proxy::plugin::dhcp::infoblox::params {
+  validate_string($username, $password)
```
Comment (iNecas) [inline manifests/plugin/dhcp/infoblox.pp]: I've noticed that the installer tried to validate theses even when the infoblox is not enabled. Is this right behaviour of the installer?

Comment (stbenjam) [inline manifests/plugin/dhcp/infoblox.pp]: It shouldn't.  Are you sure it's puppet and not Kafo? You maybe hitting https://github.com/theforeman/kafo/pull/178

Comment (ekohl) [inline manifests/plugin/dhcp/infoblox.pp]: Well, the defaults are undef which shouldn't pass as a string.

Comment (iNecas) [inline manifests/plugin/dhcp/infoblox.pp]: Then thing is, I want to force user to specify the values, having default for user credentials doesn't make much sense to me. What I think is happening is kafo validating values also for non-enabled plugins, I will try to fix this in kafo, and based on how hard the fix is, I will either do it in kafo or fix it here temporary

Comment (iNecas) [inline manifests/plugin/dhcp/infoblox.pp]: This is the kafo patch we need to make it working https://github.com/theforeman/kafo/pull/180

**Pull request closed without merging**